source-dropbox/main.go

```diff
@@ -0,0 +1,312 @@
+package main
+
+import (
+	"context"
+	_ "embed"
+	"encoding/json"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/dropbox/dropbox-sdk-go-unofficial/v6/dropbox"
+	"github.com/dropbox/dropbox-sdk-go-unofficial/v6/dropbox/files"
+	"github.com/estuary/connectors/filesource"
+	"github.com/estuary/flow/go/parser"
+	pf "github.com/estuary/flow/go/protocols/flow"
+	log "github.com/sirupsen/logrus"
+)
+
+type config struct {
+	Path        string         `json:"path"`
+	Credentials *Credentials   `json:"credentials"`
+	Parser      *parser.Config `json:"parser"`
+	MatchKeys   string         `json:"matchKeys,omitempty"`
+	Advanced    advancedConfig `json:"advanced"`
+}
+
+type advancedConfig struct {
+	AscendingKeys bool `json:"ascendingKeys,omitempty"`
+}
+
+func (c config) Validate() error {
+	requiredProperties := [][]string{
+		{"Path", c.Path},
+	}
+	for _, req := range requiredProperties {
+		if req[1] == "" {
+			return fmt.Errorf("missing '%s'", req[0])
+		}
+	}
+	return nil
+}
+
+func (c config) DiscoverRoot() string {
+	return c.Path
+}
+
+func (c config) RecommendedName() string {
+	return c.Path[1:]
+}
+
+func (c config) FilesAreMonotonic() bool {
+	return c.Advanced.AscendingKeys
+}
+
+func (c config) ParserConfig() *parser.Config {
+	return c.Parser
+}
+
+func (c config) PathRegex() string {
+	return c.MatchKeys
+}
+
+func newDropboxStore(ctx context.Context, cfg config) (*dropboxStore, error) {
+	log.Debug("Creating Dropbox client")
+	oauthClient, err := cfg.Credentials.GetOauthClient(ctx)
+	if err != nil {
+		return &dropboxStore{}, err
+	}
+	config := dropbox.Config{
+		//! Note: This Client property was originally build for testing purposes. So, it can change in the future.
+		Client:   oauthClient,
+		LogLevel: dropbox.LogDebug,
+	}
+	client := files.New(config)
+
+	store := dropboxStore{
+		client: client,
+		config: cfg,
+	}
+	if err := store.check(); err != nil {
+		return &dropboxStore{}, err
+	}
+	return &store, nil
+}
+
+type dropboxStore struct {
+	client files.Client
+	config config
+}
+
+// check checks the connection to the Azure Blob Storage container.
+// It returns an error if the container is not found, the account is disabled, or if there is an authorization failure.
+// If the listing is successful, it returns nil.
+func (dbx *dropboxStore) check() error {
+	resp, err := dbx.client.ListFolder(&files.ListFolderArg{Path: dbx.config.Path})
+	if err != nil {
+		log.Error("Failed to list files")
+		return err
+	}
+	for _, entry := range resp.Entries {
+		fileEntry, ok := entry.(*files.FileMetadata)
+		if !ok {
+			continue
+		}
+		meta, reader, err := dbx.client.Download(&files.DownloadArg{Path: fileEntry.PathDisplay})
+		if err != nil {
+			log.Error("Failed to download file: ", fileEntry.PathDisplay)
+			return err
+		}
+		log.Debug("File metadata: ", meta)
+		defer reader.Close()
+		content := make([]byte, meta.Size)
```

Comment:
This will OOM the connector if the file it happens to download is very large. We do our best to make sure the connector is configured in a way that will work, but it's not always possible to 100% verify everything. Is being able to get the metadata for a file sufficient to verify that we can read data?

Reply:
As we are using two scopes, "files.metadata.read" and "files.content.read", reading the metadata just verifies the first one. As you said, we couldn't verify 100% without acctually reading a file, so I will just check the metadata.